### PR TITLE
Add the concept of a RequestFilter

### PR DESCRIPTION
**Pull request by bryce-anderson**
A RequestFilter can be used to examine an incoming request before running the action. It returns a Task, so it can be asynchronous in its evaluation.

### CODE REVIEW

File: core/src/main/scala/org/http4s/rho/bits/ResultMatcher.scala [hunk 1]
```diff
@@ -7,10 +7,16 @@ import scala.reflect.runtime.universe.{ Type, WeakTypeTag }
 import scalaz.concurrent.Task
 
 
-trait ResultMatcher[-R] {
```
Comment (bryce-anderson) [inline core/src/main/scala/org/http4s/rho/bits/ResultMatcher.scala]: I'm thinking that `ResultMatcher` might be better as a case class with `conv` as a lambda.

**Comment by bryce-anderson** [comment]
This is on hold until #83 is addressed. It will benefit from the `Action` type, if it lands.

File: core/src/main/scala/org/http4s/rho/bits/ResultMatcher.scala (hunk 1 above)
Comment (rossabaker) [inline core/src/main/scala/org/http4s/rho/bits/ResultMatcher.scala]: And `encodings` and `resultInfo` as parameters?  Because of the `addInfo`?

**Comment by bryce-anderson** [comment]
I think this is no longer worthy of PR status. If I decided to revisit it I'll reopen with a new PR.

**Pull request closed without merging**